cmd/fil-daggregator/main.go

```diff
@@ -0,0 +1,345 @@
+	}
+
+	if err := writeoutBlocks(ctx, opts, ramBs); err != nil {
+		log.Fatalf("writing newly created dag to IPFS API failed: %s", err)
+	}
+
+	log.Infow("aggregation finished", "aggregateRoot", root, "totalEntries", len(entries))
+}
+
+func statSources(externalCtx context.Context, opts *opts, toAgg []dagaggregator.AggregateDagEntry) error {
+
+	type dagStat struct {
+		Size      uint64
+		NumBlocks uint64
+	}
+
+	ctx, shutdownWorkers := context.WithCancel(externalCtx)
+	defer shutdownWorkers()
+
+	var lastPct uint64
+	dagsDone := new(uint64)
+	dagsTotal := uint64(len(toAgg))
+
+	maxWorkers := opts.IpfsAPIMaxWorkers
+	workCh := make(chan int)                // channel of toAgg indexes to work on
+	errCh := make(chan error, 1+maxWorkers) // our own error plus one from each worker
+
+	// work dispenser, watchdog for progress and to bail early if an error appears
+	go func() {
+		defer close(workCh)
+
+		var progressTick <-chan time.Time
+		if opts.ShowProgress {
+			fmt.Fprint(os.Stderr, "0% of dags analyzed\r")
+			t := time.NewTicker(250 * time.Millisecond)
+			progressTick = t.C
+			defer t.Stop()
+		}
+
+		toAggIdx := 0
+		for toAggIdx < len(toAgg) {
+			select {
+			case workCh <- toAggIdx:
+				toAggIdx++
+			case <-ctx.Done():
+				errCh <- ctx.Err()
+				return
+			case err := <-errCh:
+				errCh <- err // put error back where we got it from
+				if err != nil {
+					shutdownWorkers()
+				}
+				return
+			case <-progressTick:
+				curPct := 100 * atomic.LoadUint64(dagsDone) / dagsTotal
+				if curPct != lastPct {
+					lastPct = curPct
+					fmt.Fprintf(os.Stderr, "%d%% of dags analyzed\r", lastPct)
+				}
+			}
+		}
+	}()
+
+	var wg sync.WaitGroup
+
+	for maxWorkers > 0 {
+		maxWorkers--
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			api := ipfsapi.NewShell(opts.IpfsAPI)
+			api.SetTimeout(time.Second * time.Duration(opts.IpfsAPITimeoutSecs))
+
+			for {
+				toAggIdx, chanOpen := <-workCh
+				if !chanOpen {
+					return
+				}
+
+				ds := new(dagStat)
+				err := api.Request("dag/stat").Arguments(toAgg[toAggIdx].RootCid.String()).Option("progress", "false").Exec(ctx, ds)
+				if err != nil {
+					errCh <- err
+					return
+				}
+
+				toAgg[toAggIdx].UniqueBlockCount = ds.NumBlocks
+				toAgg[toAggIdx].UniqueBlockCumulativeSize = ds.Size
+
+				if opts.ShowProgress {
+					atomic.AddUint64(dagsDone, 1)
+				}
+
+			}
+		}()
+	}
+
+	wg.Wait()
+	shutdownWorkers()
+
+	errCh <- externalCtx.Err()
+	return <-errCh
+}
+
+// pulls cids from an AllKeysChan and sends them concurrently via multiple workers to an API
+func writeoutBlocks(externalCtx context.Context, opts *opts, bs blockstore.Blockstore) error {
+
+	ctx, shutdownWorkers := context.WithCancel(externalCtx)
+	defer shutdownWorkers()
+
+	akc, err := bs.AllKeysChan(ctx)
+	if err != nil {
+		return err
+	}
+
+	var lastPct uint64
+	blocksDone := new(uint64)
+
+	// this works because of how AllKeysChan behaves on rambs
+	blocksTotal := uint64(len(akc))
```

Comment:
I'm sure you're aware this is not guaranteed to work in general :) e.g. the carv2 blockstore will likely use no chan buffering at all, so this would always be 0

Reply:
correct, hence the comment on top :)